src/ccs/cccs.rs

```diff
@@ -0,0 +1,279 @@
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSShape<G: Group> {
+  // Sequence of sparse MLE polynomials in s+s' variables M_MLE1, ..., M_MLEt
+  pub(crate) M_MLE: Vec<MultilinearPolynomial<G::Scalar>>,
+
+  pub(crate) ccs: CCSShape<G>,
+}
+
+/// A type that holds a CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSInstance<G: Group> {
+  // Commitment to a multilinear polynomial in s' - 1 variables
+  pub(crate) C: Commitment<G>,
+
+  // $x in F^l$
+  pub(crate) x: Vec<G::Scalar>,
+}
+
+impl<G: Group> CCCSShape<G> {
+  // Computes q(x) = \sum^q c_i * \prod_{j \in S_i} ( \sum_{y \in {0,1}^s'} M_j(x, y) * z(y) )
+  // polynomial over x
+  pub fn compute_q(&self, z: &Vec<G::Scalar>) -> Result<VirtualPolynomial<G::Scalar>, NovaError> {
+    let z_mle = dense_vec_to_mle::<G::Scalar>(self.ccs.s_prime, z);
+    if z_mle.get_num_vars() != self.ccs.s_prime {
+      // this check if redundant if dense_vec_to_mle is correct
+      return Err(NovaError::VpArith);
+    }
+
+    // Using `fold` requires to not have results inside. So we unwrap for now but
+    // a better approach is needed (we ca just keep the for loop otherwise.)
+    Ok(
+      (0..self.ccs.q).fold(VirtualPolynomial::<G::Scalar>::new(self.ccs.s), |q, idx| {
+        let mut prod = VirtualPolynomial::<G::Scalar>::new(self.ccs.s);
+
+        for &j in &self.ccs.S[idx] {
+          let sum_Mz = compute_sum_Mz::<G>(&self.M_MLE[j], &z_mle);
+
+          // Fold this sum into the running product
+          if prod.products.is_empty() {
+            // If this is the first time we are adding something to this virtual polynomial, we need to
+            // explicitly add the products using add_mle_list()
+            // XXX is this true? improve API
+            prod
+              .add_mle_list([Arc::new(sum_Mz)], G::Scalar::ONE)
+              .unwrap();
+          } else {
+            prod.mul_by_mle(Arc::new(sum_Mz), G::Scalar::ONE).unwrap();
+          }
+        }
+        // Multiply by the product by the coefficient c_i
+        prod.scalar_mul(&self.ccs.c[idx]);
+        // Add it to the running sum
+        q.add(&prod)
+      }),
+    )
+  }
+
+  /// Computes Q(x) = \hat eq(beta, x) * q(x)
+  ///               = \hat eq(beta, x) * \sum^q c_i * \prod_{j \in S_i} ( \sum_{y \in {0,1}^s'} M_j(x, y) * z(y) )
+  /// polynomial over x
+  pub fn compute_Q(
+    &self,
+    z: &Vec<G::Scalar>,
+    beta: &[G::Scalar],
+  ) -> Result<VirtualPolynomial<G::Scalar>, NovaError> {
+    let q = self.compute_q(z)?;
+    q.build_f_hat(beta)
+  }
+
+  /// Perform the check of the CCCS instance described at section 4.1
+  pub fn is_sat(
+    &self,
+    ck: &CommitmentKey<G>,
+    w: &CCSWitness<G>,
+    x: &CCCSInstance<G>,
+  ) -> Result<(), NovaError> {
+    // check that C is the commitment of w. Notice that this is not verifying a Pedersen
```

Comment:
A quick thought. I think we could potentially avoid going to the hypercube to check the satisfiability. In practice, all multilinear polynomials are going to be represented in the evaluation form, so checking the CCS vs. committed CCS could be identical. In particular, we could potentially perform the CCS satisfiability check expressed with matrix operations directly.